### PR TITLE
Enable inference mode for evaluation

### LINKED ISSUE #11018: use `torch.inference_mode()` in `Trainer.predict`
## 🚀 Feature

pytorch 1.9 introduced a new decorator/context manager specifically for model inference `torch.inference_mode()`. Like `torch.no_grad()` it disables gradient tracking for faster forward passes through a model and is, as the name implies, useful for model inference. Unlike `torch.no_grad()`, it disables view tracking, so computations made using `torch.inference_mode()` can not later be used in computations that require gradients (as opposed to `torch.no_grad()`). Disabling view tracking further speeds up model forward passes.

### Motivation

The motivation here is to implement `torch.inference_mode()` inside the `Trainer.predict()` logic, replacing any occurrences of `torch.no_grad()`. Because this code right now is contextually used for inference, I don't believe the lack of view tracking represents a significant drawback/limitation for downstream code. Conceptually, clients of `Trainer.predict()` should not be relying on these tensors for downstream code that requires gradient calculation. However, clients of this code are likely relying on pytorch lightning to transparently scale and speed up their own code, and using `torch.inference_mode()` will further this goal.

### Pitch

change
```python
def _run_predict(self) -> Optional[_PREDICT_OUTPUT]:
        self.reset_predict_dataloader(self.lightning_module)
        # reset trainer on this loop and all child loops in case user connected a custom loop
        self.predict_loop.trainer = self
        with torch.no_grad():
            return self.predict_loop.run()
```

to
```python
def _run_predict(self) -> Optional[_PREDICT_OUTPUT]:
        self.reset_predict_dataloader(self.lightning_module)
        # reset trainer on this loop and all child loops in case user connected a custom loop
        self.predict_loop.trainer = self
        with torch.inference_mode():
            return self.predict_loop.run()
```

I think this would be the only change to make

cc @borda

### LINKED ISSUE #11530: inference_mode
PyTorch has an `inference_mode` (https://pytorch.org/docs/stable/generated/torch.inference_mode.html) that it recommends as a replacement for `no_grad`. But I don't think lightning is currently using it. Should it?

cc @borda @awaelchli @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11530
Fixes #11018

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2786,6 +2788,17 @@ def configure_optimizers(self):
         return max_estimated_steps
 
 
+@contextmanager
+def _evaluation_context() -> Generator:
+    cm = (
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2793]: nit: we can spare some letters :)

```suggestion
    context_manager_class = (
```

**Review by carmocca — APPROVED** [review]
Awesome!

**Comment by ananthsub** [comment]
We previously had support for this but needed to revert it: https://github.com/PyTorchLightning/pytorch-lightning/issues/8499
https://github.com/PyTorchLightning/pytorch-lightning/pull/9443

have you confirmed this is safe to roll out now?

**Comment by rohitgr7** [comment]
> 

we have extensive tests that seem to be passing with this change, so a good indicator that this is safe. I have also confirmed that it's working with the example shared in the issue that caused the revert last time. But even if it still has some issues, we should explore the issue more in-depth rather than reverting it just like last time. I couldn't find anything related in the PyTorch docs that some backend like `gloo` is not supported or any other limitation. So I think it's safe to roll out now and explore the issue more next time if any comes up.

**Comment by rohitgr7** [comment]
> We previously had support for this but needed to revert it: #8499 #9443
> 
> have you confirmed this is safe to roll out now?

we have extensive tests that seem to be passing with this change, so a good indicator that this is safe. I have also confirmed that it's working with the example shared in the issue that caused the revert last time. But even if it still has some issues, we should explore the issue more in-depth rather than reverting it just like last time. I couldn't find anything related in the PyTorch docs that some backend like gloo is not supported or any other limitation. So I think it's safe to roll out now and explore the issue more next time if any comes up.

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py]: done!

**Review by krshrimali — APPROVED** [review]
LGTM! 🚀

**Review by akihironitta — APPROVED** [review]
LGTM! 🚀

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2744,6 +2746,17 @@ def configure_optimizers(self):
         return max_estimated_steps
 
 
+@contextmanager
+def _evaluation_context() -> Generator:
+    context_manager_class = (
+        torch.inference_mode
+        if _TORCH_GREATER_EQUAL_1_9 and not (dist.is_initialized() and dist.get_backend() == "gloo")
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:2753]: I don't find any information in docs or anywhere about inference_mode not being compatible with the gloo backend. A comment here in the code would probably be appropriate.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2744,6 +2746,17 @@ def configure_optimizers(self):
         return max_estimated_steps
 
 
+@contextmanager
+def _evaluation_context() -> Generator:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2750]:
```suggestion
def _evaluation_context() -> Generator:
    # inference mode is not supported on gloo backend
```

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2744,6 +2746,18 @@ def configure_optimizers(self):
...
+    # inference mode is not supported on gloo backend
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2751]:
```suggestion
    # inference mode is not supported with gloo backend (#9431)
```

**Review by krshrimali — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:2758]: thanks for adding the comment. i'm definitely not satisfied still, and will investigate. this is very sus, especially since I cannot find any open or closed issue ticket on the pytorch github.

Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:2758]: I couldn't find it either. Let me open an issue on PT GitHub.
